### PR TITLE
Configure Tauri as Electron alternative

### PR DESCRIPTION
**Pull request by dnaka91**
Very WIP and initial try to get Tauri up and running as alternative to Electron.

Benefits are:

- Creates a much smaller binary.
- Uses way less memory and eventually less CPU as well.

Please keep in mind that this is a two-folded experiment. On one side just from my side to finally have a proper project to try Tauri out on. On the other side really as an alternative for Electron as @negue likes to have a more lightweight solution as well.

## Current issues / TODOs

The following lists possible issues and TODOs to make the app more _ready_. Some of these might not even have been part of the Electron version and others might just be missing from meme-box itself and are better be added later as a separate feature.

Therefore, this list are just things that came up while building this initial version or a ideas for improvement.

### Depends on WebView2 on Windows

Tauri currently only bundles the app itself as installer but doesn't check for required dependencies yet. This only affects **Windows** as it needs WebView2 and the VCRedist package pre-installed. On Macos and Linux it uses the native WebView APIs.

### Interaction with the sidecar

So far it just spawns a child process and sends a signal to the script in `index.html` oncde the app is ready. That is all good but:

- Need some better readyness signal than just the text `Data saved!`.
  - There was another one `Server is ready`, but it seemed to only come up during dev mode and not in the final packaged application.
  - Would like to have something outside the logging framework, maybe some very simple JSON payload directly written to **stdout**.
- If the sidecar stops for whatever reason, the whole app will just crash with an error message as there is no good way of recovery. Could just restart the sidecar but:
  - No good way of detecting the cause of the error.
  - What if the reason is an already taken port?
  - Could create an endless loop that just restarts and restarts.
  - Probably need some signal through the Tauri API so the UI can be signalled about the error to show some error that would be forwarded from the sidecar through the Tauri app.
    - As the UI comes from the sidecar, extra care must be taken that any further action does not require the sidecar and all required components are already loaded by the browser.

### Sidecar and UI separation

For now the sidecar has everything, server backend and UI. That is great for the headless version, but for Tauri it would be better to have the UI separate from the backend. 

UI parts are already generated in the `dist/` folder but then re-bundled into the sidecar. This increases the binary size and first of all, the UI is not "aware" of Tauri yet.

In future version a separation would help to:

- Reduce binary size (as already mentioned).
- Decouple the sidecar lifecycle from the UI, allowing to handle issues with the sidecar process and handling them through a UI + Tauri integration. This solves the issue mentioned above.

### Basic OS integrations

Especially when it comes to common shortcuts, Tauri doesn't provide any defaults out of the box. That means for example on a MacOS hitting `CMD+Q` to quit or `CMD+,` to open settings doesn't have any effect. Didn't try out other things like **Copy & Paste** or Drag & Drop yet.

I'm not sure whether this is something on the roadmap for Tauri devs. So far I guess we would need to implement these by ourselves. As this needs to be done directly in Tauri, it means implementing in Rust. I understand that especially @negue doesn't have any Rust experience, so I would implement these features once we decided what shortcuts and features we want.

### Keep running in the background

This one is simple, just don't shut everything done once the application window is closed. Afaik this already works in Electron so it should be added for Tauri as well. I think Tauri even has an example for this behaviour.

Just didn't bother with it yet as I wanted to get a very first version ready.

## Other thoughts and ideas

These points are not necessarily bound to the Tauri implementation and could likely be added as separate PRs to improve some minor aspects of meme-box.

- Currently the sidecar fails if a port is already taken. For the headless version that's fine but for the Tauri version it would be great if the sidecar would just pick a random port until it finds one (most OS APIs already have this feature by setting the port to `0` which lets the kernel decide the port). The resulting port could be send to Tauri together with the readyness signal as JSON as outlined further above.
- Allow to define the network interface and default to `localhost/127.0.0.1`. Currently memebox seems to use `0.0.0.0` by default and doesn't have a way of changing that. It's great if you want to connect your phone through the respective feature but also a security issue (as there is no auth so far). When opening the app instead of the headless version, this effect is even less clear as it might not be in the understanding of every user that a server is running in the background. And suddenly having this UI + server available to everyone on the same network can be risky (think using it from an internet cafe, general dangers of connecting to a public WiFi aside).

### CODE REVIEW

**Comment by negue** [comment]
WOW - I've been keeping an eye on Tauri and its hopefully soon upcoming "Deno Bindings"
for this PR - I still need to try it / read it fully :D but I can answer 1-2 things:

> Currently the sidecar fails if a port is already taken. For the headless version that's fine but for the Tauri version it would be great if the sidecar would just pick a random port until it finds one (most OS APIs already have this feature by setting the port to 0 which lets the kernel decide the port). The resulting port could be send to Tauri together with the readyness signal as JSON as outlined further above.

A fixed port is needed since all browersources use that inside OBS - so I dont think this will be ever changed / "fixed"

But some kind of events to Tauri could be definitely added, is there an example of how other sidecar apps handle the communication ?

> Allow to define the network interface and default to localhost/127.0.0.1. Currently memebox seems to use 0.0.0.0 by default and doesn't have a way of changing that. It's great if you want to connect your phone through the respective feature but also a security issue (as there is no auth so far). When opening the app instead of the headless version, this effect is even less clear as it might not be in the understanding of every user that a server is running in the background. And suddenly having this UI + server available to everyone on the same network can be risky (think using it from an internet cafe, general dangers of connecting to a public WiFi aside).

An ability to change the listening IP should be added yes. It was 0.0.0.0 so that everyone can use the mobile view to trigger stuff. When there is an ability to change the IP then ppl would have to decice - for mobile view / remote access or localhost only  (sure on localhost the mobile view would work too)



I think any other ideas / questions we could move those to the Discussions Feature so that this PR won't be  spammed :D

**Comment by negue** [comment]
could you add the basic builds to the github workflow ?  (probably with an variable to enable those sub builds, examples in the workflow)

also how about (if I) change the target branch to be one for tauri specific? then the workflows and changes would still be separated until a first versin would be able to added to all nightly builds

File: projects/action-variables-ui/package-lock.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (negue) [inline projects/action-variables-ui/package-lock.json]: those files dont even need to be installed - so a packagelog doesn't even make sense here :P

File: scripts/move-binary.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/**
+ * This script is used to rename the binary with the platform specific postfix.
+ * When `tauri build` is ran, it looks for the binary name appended with the platform specific postfix.
+ */
+
+const execa = require('execa')
```
Comment (negue) [inline scripts/move-binary.js:6]: I think this can be replaced by the built-in `child_process` module and the `exec` function

File: src-tauri/index.html [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        justify-content: center;
+      }
+    </style>
+  </head>
+
+  <body>
+    <div>Starting up...</div>
+    <script>
+      window.__TAURI__.event.listen('ready', (event) => {
+        window.location.href = 'http://localhost:6363'
```
Comment (negue) [inline src-tauri/index.html]: this would still need the custom port if it was set from the server

File: .github/workflows/build_and_publish.yml [hunk 4]
```diff
@@ -47,9 +47,9 @@ jobs:
             **/node_modules
           key: ${{ runner.os }}-${{ hashFiles('**/package.json') }}
       - name: Use Node.js 12.16.2
-        uses: actions/setup-node@v1
+        uses: actions/setup-node@v2
         with:
-          node-version: 12.16.2
+          node-version-file: .nvmrc
```
Comment (negue) [inline .github/workflows/build_and_publish.yml]: nice way with the .nvmrc 👍

File: package.json [hunk 5]
```diff
@@ -54,7 +54,14 @@
     "electron:build:serve": "npm run electron:build_0compile && npm run electron:build_1combineJs &&  npx electron out-electron/index.js --serve --built-serve",
     "docs:json": "compodoc -p ./tsconfig.base.json -e json -d .",
     "storybook": "npm run docs:json && start-storybook -p 6006",
-    "build-storybook": "npm run docs:json && build-storybook"
+    "build-storybook": "npm run docs:json && build-storybook",
+    "package:windows": "npm run build:prepare && npm run build:windows && node scripts/move-binary.js",
```
Comment (negue) [inline package.json]: since `build:prepare` and `build:<OS>` is running before the tauri builds anyway you don't need to call em again - that'll save up a good chunk of each pipelines time

**Review by negue — COMMENTED** [review]
Added some notes / hints to the changes

in #417 I added a cli-option to let the `stdout`  be json instead that could help with the port info issue you mentioned

could you try if its possible to use the same images/icons that already exist? then there wouldn't be need to have it duplicated

also could you try to run the pipeline on your repo as well? seems that the current builds couldn't find any rust built executables  (or those didn't build and the pipeline didn't report it as an error)

File: scripts/move-binary.js (hunk 2 above)
Comment (dnaka91) [reply scripts/move-binary.js:6]: Probably 🤷
This is just a slightly adjusted version from the Tauri docs.

File: src-tauri/index.html (hunk 3 above)
Comment (dnaka91) [reply src-tauri/index.html]: One step at a time. So far the Tauri version doesn't support to pass additional arguments yet.

When that's added, we can of course pass the port as event parameter and use it here to apply the right port.

File: .github/workflows/pullrequest.yml [hunk 6]
```diff
@@ -34,35 +33,34 @@ jobs:
...
       - uses: actions/upload-artifact@v2
         with:
           name: ${{ matrix.os }}_output
-          path: |
-            release/out/
-            release-electron/
-            !release/config/**
-            !release-electron/*-unpacked/**
-            !release-electron/mac/**
+          path: src-tauri/target/release/bundle/**/*
```
Comment (negue) [inline .github/workflows/pullrequest.yml]: `release/out/`  would be still needed for the headless build

File: server/server-app.ts [hunk 7]
```diff
@@ -56,7 +56,7 @@ PersistenceInstance.hardRefresh$()
     startWith(true)
   )
   .subscribe(() => {
-    console.info('Data Hard-Refresh');
+    LOGGER.info('Data Hard-Refresh');
```
Comment (negue) [inline server/server-app.ts]: oh wow you replaced a bunch of loggers - those could be its own PR (if you have the time for it - if not - no worries here :)

File: .github/workflows/pullrequest.yml [hunk 8]
```diff
@@ -34,35 +33,34 @@ jobs:
...
+        uses: Swatinem/rust-cache@v1
       - run: npm install
-      - run: npm run build:prepare
-      - run: npm run build:nexe -- -o release/out/memebox
-
-      - run: release/out/memebox --cli-test-mode=true  --config=./release/config
-
-      - run: npm run electron:build:only
-
+      - run: npm build:prepare
```
Comment (negue) [inline .github/workflows/pullrequest.yml]: npm run build:prepare ^^ 

probably would be easier if you try it on your repo / pr (to it)

**Review by negue — COMMENTED** [review]
some hints - tried to run the pipeline sadly didn't work yet

File: .github/workflows/pullrequest.yml (hunk 8 above)
Comment (dnaka91) [reply .github/workflows/pullrequest.yml]: Just did that yesterday evening.
I'll redo the CI tasks altogether, as I want to re-add the Electron builds as well.

File: server/server-app.ts (hunk 7 above)
Comment (dnaka91) [reply server/server-app.ts]: By the end of it, I figured that 😆. It's just so we can handle pretty much all the logs properly, or otherwise we'd have a mix of JSON logs and plain text logs.

Maybe we can add something to `eslint` to deny any `console` calls in the `server/**` files?

Also, what happens to other calls to `console` outside the server package (that are shared between UI and server, if any)?
Perhaps we move the logger into a separate package that all packages depend on with a common interface. Then the UI forwards to console calls and the server forwards to the `NamedLogger`. What do you think?

Comment (negue) [reply server/server-app.ts]: right before this PR there wasn't a need (yet, it was on my todo list yes) to have all logs done in a normed way

I can try to add some lints for those yeah, as for a shared way (since some libs are used between server and UI) I cant use the one from TsED ^^  soo we'd need some kind of shared logger 

all TsED Services/Controller get it by using dependency injection - all others would need some kind of interface which is set by each platform (server, web) - I don't expect that to happen fully today but the linting part will

File: .github/workflows/pullrequest.yml [hunk 9]
```diff
@@ -2,67 +2,144 @@ name: Pull-Request CI
...
         uses: actions/cache@v2
         with:
-          path: |
-            **/node_modules
-          key: ${{ runner.os }}-${{ hashFiles('package.json') }}
-      - name: Use Node.js 12.16.2
-        uses: actions/setup-node@v1
+          path: node_modules/.cache
+          key: pr-angular-${{ runner.os }}-${{ hashFiles('**/package-lock.json') }}
+      - run: npm install
```
Comment (negue) [inline .github/workflows/pullrequest.yml]: the reason I had it all under one build group is that I didn't wanted to waste the time for
- `npm install`
- `npm build:prepare` 

each time, which installs it all, compiles angular stuff, compiles the basic server files and thats between 6-10 minutes

there are ways to share the output BUT I need try that again --- I think I had it done for the release pipeline

File: .github/workflows/pullrequest.yml [hunk 10]
```diff
@@ -2,67 +2,144 @@ name: Pull-Request CI
...
-
-      - uses: actions/upload-artifact@v2
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: electron-${{ matrix.os }}
+          path: |
+            release-electron/memebox*.dmg
+            release-electron/memebox*.AppImage
+            release-electron/memebox*.exe
```
Comment (negue) [inline .github/workflows/pullrequest.yml]: I think those are under `/release-electron/out`

File: .github/workflows/pullrequest.yml (hunk 9 above)
Comment (dnaka91) [reply .github/workflows/pullrequest.yml]: I hoped splitting them up would speed up things. But the majority of time is spent in those two steps you mentioned.

Will move those back together again 👍 

What probably works is, to run `install` and `build:prepare`, then upload `node_modules` and the needed directories as artifact. In the following jobs, download said artifact into the current folder before anything else.

Comment (negue) [reply .github/workflows/pullrequest.yml]: yeah the node_modules probably not helping in terms of speed - those could stay in each build for now

but everything else - the built angular app and the built server output that can be shared betweet them all

also the `prepare:` don't need OS dependency - the created files are js nothing native there

File: server/server-app.ts (hunk 7 above)
Comment (negue) [reply server/server-app.ts]: speaking of "Done" Logger Entry

wouldn't this one work? 

`Server is ready`  ?

**Comment by negue** [comment]
woops I misclicked and " marked this pull request as ready for review"  ... sorry :/

**Comment by negue** [comment]
I hope to get this review done by the Weekend - sorry for the LONG wait

**Review by negue — APPROVED** [review]
Sorry it took that long - real life and what not

Looking good - lets see if the current CI works out and then this can be merged in :>

**Comment by dnaka91** [comment]
Hm seems the `Run Affected Lint` step still fails at the file `server/providers/websockets/memebox.websocket.ts`, even though the PR doesn't touch that file 🤔

**Comment by negue** [comment]
It seems some of the changes increase the linting level =D  - see/copy my latest two commits in https://github.com/negue/meme-box/pull/552  - then the build/lint works

While running the CI - is there a way to cache some of the rust stuff? 
it took now 15min to build and I think it'll build the same-ish stuff most of the time (if I dont change the tauri code) - so maybe there is cache there? if not then so be it xD

**Comment by dnaka91** [comment]
And here I was so certain, I didn't change anything about the other files :sweat_smile:. Fingers crossed that the CI passes now.

The CI includes a cache step for the Rust compilation. Therefore, the first run might be quite slow, but afterward it should compile a lot faster, as long as the Cargo.toml/Cargo.lock files don't change.

**Comment by negue** [comment]
Thank you for this HUGE amount of work and tries and errors and repeats haha - the next release will have a beta version of the Tauri executable as well  - or the daily today - we'll see if it all still works -- Vercel build somehow broke... why? dunno yet =D

Anyway - thank you!! this is big
